### ISSUE TITLE
Error messages are less than helpful

### ISSUE THREAD

**Issue by taralx** [alexcrichton/tar-rs#128]
"numeric field was not a number" doesn't tell you *which* field is not a number, or in which of the many parts of the tar file the error occurred. It's only barely better than just saying "bad" with no other detail.

**Comment by alexcrichton** [comment]
Indeed! It'd be great to do a pass over the error strings and make them more descriptive where possible

**Comment by vitiral** [comment]
I am including this library as part of [ergo_fs](https://docs.rs/ergo_fs/0.1.0/ergo_fs/#tar-files) and have similar requirements for improved error messages.

Primarily I would like to include all of the context of what went wrong on _any_ error message. This especially includes the `path` information for any operations that fail on the filesystem.

General strategy:
- For packing files from the filesystem, simply do a `map_err` that includes the context in the `io::Error`. Rely on the `Write` object to provide it's own context (i.e. [FileWrite](https://docs.rs/ergo_fs/struct.FileWrite.html) will automatically include context on any failed `write`).
- When unpacking files, simply a `map_err` for any failed writes to the filesystem. Rely on the `Read` object to provide it's own context (similar to `FileWrite`, `FileRead` in the same module provides context for errors)

If you approve I can begin work on this shortly.

**Comment by alexcrichton** [comment]
@vitiral sounds great to me! I'd also be ok adding a new error type to this library but sticking to `io::Error` I think is fine for now.

**Comment by vitiral** [comment]
@alexcrichton sounds good!

**Comment by vitiral** [comment]
@taralax you're issue was a _teensy_ bit more tricky than I had thought, but I think I got it figured out 😄
